Update KiNET plugin README for Port out (#40)
peternewman/ola: plugins/kinet/README.md

============

This plugin creates a single device with multiple output ports. Each port

> represents a power supply. This plugin supports both the V1 DMX out and the
> V2 Port out modes of the KiNET protocol.

Reviewer: They seem to style it as this:
```suggestion
represents a power supply. This plugin supports both the V1 DMX Out and the
V2 Port Out modes of the KiNET protocol.
```

Well actually they go for DMX OUT too, but it looks a bit odd to me.

Although as per the other discussion I think we should probably go more like:
```suggestion
represents a power supply. This plugin supports both the DMX Out (V1) and the
Port Out (V2) modes of the KiNET protocol.
```

Author: Fixed.

## Config file: `ola-kinet.conf`

`power_supply = <ip>`
The IP of the power supply to send to. You can communicate with more than
one power supply by adding multiple `power_supply =` lines.

> `<ip>-mode = [v1dmxout|v2portout]`
> The mode of KiNET to send to the power supply.

Reviewer: As mentioned elsewhere:
```suggestion
`<ip>-mode = [dmxout|portout]`
The mode of KiNET to send to the power supply. DMX Out is sometimes known as V1 and Port Out as V2.
```

Author: Fixed.

In general, it is highly recommended to use the V2 Port out mode if your power
supply supports it. Note that some newer devices only support the V2 Port out
protocol. On some devices it is also necessary to use the V2 Port out protocol
in order to address all of the connected fixtures if the number of channels
required exceeds one universe.

Note that for DMX out mode, the universe option on the power supply will be
ignored, as this plugin supports only unicast DMX out packets. Instead, the
universe may be patched by assigning this output port to the intended universe

> in OLA.

Reviewer: Should we explain that OLA sends to the KiNET broadcast universe or list the actual value?

Author: I added a note about the broadcast. In practice, I am not sure that the exact value will be relevant to most users.

============

This plugin creates a single device with multiple output ports. Each port
represents a power supply. This plugin supports both the DMX Out (V1) and the
Port Out (V2) modes of the KiNET protocol.

## Config file: `ola-kinet.conf`

### Per Power Supply Settings

`power_supply = <ip>`
The IP of the power supply to send to. You can communicate with more than

> one power supply by adding multiple `power_supply =` lines.

Reviewer: I think this should probably be above the sub heading but in the config file bit, as it's not a setting of the PSU but something that makes them exist. Which would be more relevant if we add auto-detection.

Author: Moved.

`<ip>-mode = [dmxout|portout]`
The mode of KiNET to send to the power supply. DMX Out is sometimes known as V1
and Port Out as V2.

> In general, it is highly recommended to use the Port Out mode if your power
> supply supports it. Note that some newer devices only support the Port Out
> protocol. On some devices it is also necessary to use the Port Out protocol in

Reviewer: ```suggestion
In general, it is highly recommended to use the Port Out mode if your power
supply supports it, we default to DMX Out mode for backwards compatibility reasons. Note that some newer devices only support the Port Out
protocol. On some devices it is also necessary to use the Port Out protocol in
```

> Note that some newer devices only support the Port Out protocol.

As an end user, how can I tell? In QPP, or do I jut have to RTFM?

Author: RTFM (if it is CK they may not even put it there), or to be honest just trial and error. Considering the oldest device I have used from 2005 supports Port Out, it shouldn't be a surprise really. Though if the device supports more channels than can fit in one universe (like the sPDS-480ca), then it is very likely.

Author: Added backwards compatibility note.

Reviewer: > RTFM (if it is CK they may not even put it there), or to be honest just trial and error. Considering the oldest device I have used from 2005 supports Port Out, it shouldn't be a surprise really. Though if the device supports more channels than can fit in one universe (like the sPDS-480ca), then it is very likely.

I wonder if we should be more positive/pro-active with a try it and see note, although I guess it's only relevant if its driving more than a universe in which case it presumably probably does anyway?

Author: Yeah it does already say that devices that require more channels will need it. I can add a note that Port Out should be tried first.

Author: Added note to try it first. There are already notes about the newer devices and the larger-than-one-universe.

order to address all of the connected fixtures if the number of channels
required exceeds one universe.

Note that for the DMX Out mode the universe option on the power supply will be

> ignored, as this plugin supports only unicast DMX Out packets destined for the

Reviewer: Grammar:
```suggestion
ignored, as this plugin only supports unicast DMX Out packets destined for the
```

Author: Fixed.

> broadcast universe on each device. Instead, the universe for each device may be

Reviewer: Do they call it broadcast? Would wildcard or catch-all be less confusing with mixiing terms?

Author: Fixed.

`power_supply = <ip>`
The IP of the power supply to send to. You can communicate with more than
one power supply by adding multiple `power_supply =` lines.

### Per Power Supply Settings

`<ip>-mode = [dmxout|portout]`
The mode of KiNET to send to the power supply. DMX Out is sometimes known as V1
and Port Out as V2.

In general, it is highly recommended to use the Port Out mode if your power
supply supports it. OLA defaults to DMX Out mode for backwards compatibility
reasons. Note that some newer devices only support the Port Out protocol. On
some devices it is also necessary to use the Port Out protocol in order to
address all of the connected fixtures if the number of channels required
exceeds one universe.

Note that for the DMX Out mode the universe option on the power supply will be
ignored, as this plugin only supports unicast DMX Out packets destined for the
wildcard universe on each device. Instead, the universe for each device may be
patched by assigning this output port to the intended universe in OLA.

`<ip>-ports = <int>`
The number of physical ports available on the power supply in Port Out mode.
Each physical port will create an OLA port that may be assigned to any

> universe. This setting is ignored in DMX Out mode.

Reviewer: Should we maybe add that it defaults to 16 which is the max?

Author: Added.

## Config file: `ola-kinet.conf`

`power_supply = <ip>`
The IP of the power supply to send to. You can communicate with more than
one power supply by adding multiple `power_supply =` lines.

### Per Power Supply Settings

`<ip>-mode = [dmxout|portout]`
The mode of KiNET to send to the power supply. DMX Out is sometimes known as
V1 and Port Out as V2.

In general, it is highly recommended to use the Port Out mode if your power
supply supports it. Only very old devices do not support Port Out mode, so
it is recommended to try Port Out mode first. OLA defaults to DMX Out mode
for backwards compatibility reasons. Note that some newer devices only
support the Port Out protocol. On some devices it is also necessary to use
the Port Out protocol in order to address all of the connected fixtures if
the number of channels required exceeds one universe.

Note that for the DMX Out mode the universe option on the power supply will
be ignored, as this plugin only supports unicast DMX Out packets destined
for the wildcard universe on each device. Instead, the universe for each
device may be patched by assigning this output port to the intended universe

> in OLA.

Reviewer: A bit retrospective, but do you think this would be clearer @DaAwesomeP ?

`Instead, each device may be patched by assigning this output port to the intended OLA universe.`

Or something similar, just to try and avoid people thinking the OLA universe has any bearing on KiNet land, or indeed matters at all...

`<ip>-ports = <int>`
The number of physical ports available on the power supply in Port Out mode.
Each physical port will create an OLA port that may be assigned to any

> universe. This setting is ignored in DMX Out mode. The default and maximum
> number of ports per power supply supported by OLA is 16.

Reviewer: ```suggestion
universe. This setting is ignored in DMX Out mode. The default and maximum
number of ports per power supply is 16.
```

The external tool I used only allowed up to 16 ports, should I allow/generate a larger range then?

Author: Technically based on the fact that it is a whole byte it should be able to go up to 255, though I have never encountered one with more than 8 ports. I think a more reasonable default is 2 ports (most common across all CK devices a least). Most devices I have encountered are either 1, 2, 4, or 8 ports.

Reviewer: > Technically based on the fact that it is a whole byte it should be able to go up to 255, though I have never encountered one with more than 8 ports.

Yeah, given the non-CK tool I tried was limited to 16, I assumed that might be a formal limit in the protocol.

> I think a more reasonable default is 2 ports (most common across all CK devices a least). Most devices I have encountered are either 1, 2, 4, or 8 ports.

My main reason for default to the max is you then don't need to change the setting, you can just avoid patching the non-existent ports, although the option is present should you want to tidy things up.

Ideally we'd detect and auto-configure per device but I've no idea if that's actually possible.

Author: OK, then let's stick with 16.

>Yeah, given the non-CK tool I tried was limited to 16, I assumed that might be a formal limit in the protocol.

I just checked QuickPlay Pro 2 and it also only lets you add a device with 16, so I guess that must be the maximum.
